### PR TITLE
Public calendar

### PR DESCRIPTION
**Pull request by matthiasmoritz**
This Feature allows users to provid the own Calendar to all (and even anonymus) visitors.

It's my first pice of code for friendica, and my first pull request to a open source project so code quality might be not perfect ;)

### CODE REVIEW

**Comment by matthiasmoritz** [comment]
By the way, is there any tutorial how to execute the checks?

File: src/Model/Event.php [hunk 1]
```diff
@@ -557,18 +558,38 @@ public static function getByIdAndUid(int $owner_uid, int $event_id, string $nick
...
+
+	}
+
+
+	public static function getPublicById( int $event_id ): array
+	{
+		/* To prevent from requesting private events via api we must add something like
+		 * $sql_perms = $sql_extra = and allow_cid is null and allow_gid is null and deny_cid is null and deny_gid is null
+		 * but I guess there is a methode alredy implemented for this?
+		 */
```
Comment (nupplaphil) [inline src/Model/Event.php]: Please move this part to a PHP Doc header

File: src/Model/Event.php [hunk 2]
```diff
@@ -557,18 +558,38 @@ public static function getByIdAndUid(int $owner_uid, int $event_id, string $nick
...
+		$events = self::getById( $event_id, '' );
```
Comment (nupplaphil) [inline src/Model/Event.php]:
```suggestion
		$events = self::getById($event_id, '');
```

File: src/Model/Event.php [hunk 3]
```diff
@@ -557,18 +558,38 @@ public static function getByIdAndUid(int $owner_uid, int $event_id, string $nick
...
+		return $events[0];
+	}
+
+	private static function getById ( int $event_id, string $additionalCondition) : array
```
Comment (nupplaphil) [inline src/Model/Event.php]:
```suggestion
	private static function getById (int $event_id, string $additionalCondition) : array
```

File: src/Module/Calendar/Event/Get.php [hunk 4]
```diff
@@ -52,18 +54,23 @@ public function __construct(L10n $l10n, App\BaseURL $baseUrl, App\Arguments $arg
...
+			// get events by id or by date
+			if (!empty($request['id']) && !Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
+				$events = [Event::getByIdAndUid($this->session->getLocalUserId(), $request['id'], $this->parameters['nickname'] ?? null)];
+			} elseif (!empty($request['id']) && Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
+				$events = [Event::getPublicById($request['id'])];
+			} else {
+				$events = Event::getListByDate($this->session->getLocalUserId(), $request['start'] ?? '', $request['end'] ?? '', false, $this->parameters['nickname'] ?? null);
+			}
+	
+			System::jsonExit($events ? self::map($events) : []);
```
Comment (nupplaphil) [inline src/Module/Calendar/Event/Get.php]:
```suggestion
		if (!$this->session->getLocalUserId() && !Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
			throw new HTTPException\UnauthorizedException();
		}

		// get events by id or by date
		if (!empty($request['id']) && !Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
			$events = [Event::getByIdAndUid($this->session->getLocalUserId(), $request['id'], $this->parameters['nickname'] ?? null)];
		} elseif (!empty($request['id']) && Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar')) {
			$events = [Event::getPublicById($request['id'])];
		} else {
			$events = Event::getListByDate($this->session->getLocalUserId(), $request['start'] ?? '', $request['end'] ?? '', false, $this->parameters['nickname'] ?? null);
		}

		System::jsonExit($events ? self::map($events) : []);
```

File: src/Module/Calendar/Event/Get.php [hunk 5]
```diff
@@ -52,18 +54,23 @@ public function __construct(L10n $l10n, App\BaseURL $baseUrl, App\Arguments $arg
 
 	protected function rawContent(array $request = [])
 	{
-		if (!$this->session->getLocalUserId()) {
-			throw new HTTPException\UnauthorizedException();
-		}
+		$calenderOwner= User::getOwnerDataByNick(!empty($this->parameters['nickname']) ? $this->parameters['nickname'] : '');
+		//if (Feature::isEnabled($calenderOwner['uid'] ?? $this->session->getLocalUserId(), 'public_calendar') && !empty($request['id']) ){
```
Comment (nupplaphil) [inline src/Module/Calendar/Event/Get.php]: Can you delete this if it isn't necessary anymore?

**Review by nupplaphil — CHANGES_REQUESTED** [review]

**Comment by MrPetovan** [comment]
The checks are meant to be run by GitHub when you submit a PR and any time you push the branch. If you want to perform these checks in advance of the review, you can open a Draft PR (or convert this one to Draft) so that there's no pressure to add more changes to it.

**Comment by nupplaphil** [comment]
Thank you for your help!! :)

Additionally, can you:
- Rebase this branch onto the latest `develop`. This will fix some checks
- Recreate the `view/lang/C/messages.po` --> plz execute the `bin/run_xgettext.sh` script to do that

**Comment by MrPetovan** [comment]
Thank you for your work! For consistency, Friendica currently enforces the PSR-2 coding standards, see https://www.php-fig.org/psr/psr-2/

I've noticed too many violations in your changes to individually point them out, would like to take a first pass or for me to take over the branch?

**Comment by matthiasmoritz** [comment]
@MrPetovan let me first go through the code. There is also still one functional issue that can cause privacy errors. I'll update the PR.

**Comment by matthiasmoritz** [comment]
@MrPetovan can you please take over the PR. The Feature is working as excepted but I'm not shure how to continiue with the rest.

**Comment by MrPetovan** [comment]
I will, thank you for your work!

**Comment by MrPetovan** [comment]
Oh Lord, I'm diving down the calendar rabbit hole and nothing about the whole feature seems to make sense (not your specific changes) and I'm rewriting way more than I should but this is the life that chose me.

**Comment by matthiasmoritz** [comment]
Sorry for opening pandoras box 😬

**Review by Quix0r — APPROVED** [review]
Type-hint checks are fine so far.

**Comment by MrPetovan** [comment]
Alright! This took me on two separate tangents (the theme variables and the `hidewall` setting) before getting to the main issue which is event retrieval and permission check, but it's there.

@matthiasmoritz You may be surprised to find very little of your own changes in the final PR, this is because I've simplified the existing `Model\Event` methods, which made many of your changes unnecessary.

Overall, there was a lot of unused/confusing code I removed and it's always a good thing.

The `hidewall` setting has been normalized (it was inconsistent in checking for authentication) and now covers a few more endpoints, including syndication feeds, so I've edited the setting description to reflect this new state of affairs, hopefully it's clearer.

**Review by Quix0r — APPROVED** [review]
Type-hint check okay.

**Comment by nupplaphil** [comment]
Here we go :-) LGTM

**Comment by MrPetovan** [comment]
@matthiasmoritz Thank you for your work again! You've helped improve Friendica on three separate fronts!

**Comment by matthiasmoritz** [comment]
@MrPetovan  Not so sure how my three lines of code contributed to your work :) but anyway, I can test the feature and i think you broke something in the UI. See #12308
image

**Comment by MrPetovan** [comment]
Are you sure it was working before? Do you want to look into it? This display comes from the `Module\Event\Show` class.

I've noticed we were performing a `Strings::escapeHTML` before a `BBCode::convert` so this may be the result of a double-encoding issue.

**Comment by matthiasmoritz** [comment]
It works on stable 🤷‍♂️

**Comment by MrPetovan** [comment]
So it could be a regression. I'll look into it if you don't want to.

**Pull request merged**